crates/volta-migrate/src/v1.rs

```diff
@@ -0,0 +1,105 @@
+use std::convert::TryFrom;
+#[cfg(unix)]
+use std::fs::remove_file;
+use std::fs::File;
+use std::path::PathBuf;
+
+use super::empty::Empty;
+use super::regenerate_shims_for_dir;
+use super::v0::V0;
+use log::debug;
+use volta_core::error::ErrorDetails;
+#[cfg(unix)]
+use volta_core::fs::read_dir_eager;
+use volta_fail::{Fallible, ResultExt, VoltaError};
+use volta_layout::v1;
+
+pub struct V1 {
+    pub home: v1::VoltaHome,
+}
+
+impl V1 {
+    pub fn new(home: PathBuf) -> Self {
+        V1 {
+            home: v1::VoltaHome::new(home),
+        }
+    }
+
+    pub fn finalize(&self) -> Fallible<()> {
+        regenerate_shims_for_dir(self.home.shim_dir())?;
+
+        debug!("Writing layout marker file");
+        File::create(self.home.layout_file()).with_context(|_| {
+            ErrorDetails::CreateLayoutFileError {
+                file: self.home.layout_file().to_owned(),
+            }
+        })?;
+
+        Ok(())
+    }
```

Comment:
The name `finalize` here makes me think of this in terms of a builder-style item; that in turn makes me wonder if this should be `Fallible<&Self>`? Probably a premature refactoring given I can't actually think of what you'd do with it *afterward*, unless maybe something like `some_v1.finalize().home` or something.

A related, but less trivial, question I have here: is it valid to use a `V1` *without* its being finalized? If it's *not*, what's your thought on *actually* having a builder so that it's impossible to end up with a type that *seems* like everything is good to go, but where it's out of sync with the state of the file system?

Reply:
`finalize` is probably a poor name, maybe `commit()` would be better? The operation here is the one that essentially marks the migration as "completed" by writing the layout file and updating the shims to their new target. Once this completes, any future migrations will see the layout file and work under the assumption that the file system is laid out as V1 expects. That's why it's the last thing to happen, to make sure that a failure anywhere along the migration process will allow it to re-start without issue.

Reply:
Based on feedback from @dherman and thinking through the flow a little more, I removed this method entirely, so the naming isn't a problem!